### PR TITLE
Bulkimport

### LINKED ISSUE #11081: User bulk import does not prevent case insensitive username matches
## Observed behavior
Case insensitive username matching during user creation is enforced in the FacilityUserManager class create_user function: https://github.com/learningequality/kolibri/blob/release-v0.16.x/kolibri/core/auth/models.py#L616

However, the bulk import code only checks if there is a username clash if a UUID has been specified for the user, and the user UUID already exists - otherwise, it is not checked. https://github.com/learningequality/kolibri/blob/release-v0.16.x/kolibri/core/auth/management/commands/bulkimportusers.py#L559

Also, when the duplicate username check is made, it is not done using a case insensitive check: https://github.com/learningequality/kolibri/blob/release-v0.16.x/kolibri/core/auth/management/commands/bulkimportusers.py#L567

Lastly, it might be useful to validate that the CSV itself does not contain case insensitive username duplicates.

### PR DESCRIPTION
**Pull request by nick2432**
## Summary
Description of the Change:
The proposed changes aim to enhance the user creation process in the FacilityUserManager class by addressing issues related to case-insensitive username matching and the verification of duplicate usernames during bulk user imports.
Case-Insensitive Username Matching (FacilityUserManager):

In the create_user function of the FacilityUserManager class, case sensitivity is introduced to the username matching process. This ensures that usernames are checked in a case-sensitive manner during user creation.
Link to the relevant code: [FacilityUserManager]
Bulk User Import Duplicate Username Check (bulkimportusers.py):
Previously, the bulk import code only checked for username clashes if a UUID was specified for the user. The proposed changes extend this check to cases where the UUID is not specified, providing a more comprehensive verification process.
…
## References
Fixes https://github.com/learningequality/kolibri/issues/11081

…

### CODE REVIEW

File: kolibri/core/auth/models.py [hunk 1]
```diff
@@ -639,7 +639,7 @@ def create_user(self, username, email=None, password=None, **extra_fields):
         if extra_fields.get("facility") is None:
             extra_fields["facility"] = Facility.get_default_facility()
         if self.filter(
-            username__iexact=username, facility=extra_fields["facility"]
+            username=username, facility=extra_fields["facility"]
```
Comment (rtibbles) [inline kolibri/core/auth/models.py]: This should not be updated - the intention was for the bulk import flow to match the enforcement that happens here, not for this enforcement to be removed.

File: kolibri/core/auth/management/commands/bulkimportusers.py [hunk 2]
```diff
@@ -579,6 +579,21 @@ def build_users_objects(self, users):
                 if self.compare_fields(user_obj, values):
                     update_users.append(user_obj)
             else:
+                # If UUID is not specified, check for a username clash
+                if values["uuid"] == "":
+                    existing_user = FacilityUser.objects.filter(
+                        username=user, facility=self.default_facility
```
Comment (rtibbles) [inline kolibri/core/auth/management/commands/bulkimportusers.py]: We should do a case insensitive match here as well, to parallel the changes you made above.

**Review by rtibbles — CHANGES_REQUESTED** [review]
2 changes needed, to have consistent case insensitivity.

**Comment by nick2432** [comment]
@rtibbles  I have changed what you said.

**Review by rtibbles — COMMENTED** [review]
Excellent, this looks good. We'll have manual QA test this, but I think this should be mergeable soon.

We will hold off on merging it until the 0.16.0 release is out.

In the meantime, could you address the linting issue? For now we can ignore the complexity requirement by doing this:

```
def build_users_objects(self, users):  # noqa C901
```

This is a directive comment to our linter to ignore this rule for this function for now.

**Comment by nick2432** [comment]
@rtibbles , do you have an estimate for when this will be merged?

**Comment by rtibbles** [comment]
After 0.16.0 is released! We're in the final stages of QA now, and made a release candidate yesterday, so it is close :)

**Comment by pcenov** [comment]
Hi @nick2432 and @rtibbles - I confirm that it's no longer possible to import a duplicate username (regardless of the letter case) if it's already extant in the facility.
However if there are several duplicate usernames in the CSV file (with different letter case like Peter, PETER) and I am importing them for the first time it's still possible to import all of them. So the question is whether this is in scope for this PR and should be fixed as well?

https://github.com/learningequality/kolibri/assets/79847249/5d249098-56fe-47db-a369-c3753d749877

**Comment by rtibbles** [comment]
That's a good catch, Peter - I think we can file that as a follow up issue for now, so as not to block @nick2432 further here, but very happy for him to pick it up if he's interested!

**Comment by nick2432** [comment]
Thank you for the clarification, @rtibbles . I'll proceed with the current PR as suggested. I'll make a note of the additional issue with importing several duplicate usernames and will file it as a follow-up task. If there are any specific considerations or details I should be aware of while working on the current PR, please let me know. Appreciate your guidance!

**Comment by pcenov** [comment]
Hi @nick2432 I have filed the follow-up issue here: https://github.com/learningequality/kolibri/issues/11890
I didn't observe any other issues while manually testing so if you are not going to be making any other changes here we could go ahead and approve this PR.

**Comment by nick2432** [comment]
> Hi @nick2432 I have filed the follow-up issue here: #11890 I didn't observe any other issues while manually testing so if you are not going to be making any other changes here we could go ahead and approve this PR.

ok

**Review by radinamatic — APPROVED** [review]
Manual QA passed, with one follow-up issue opened, good to go! 💯 🚀 :shipit:

**Review by rtibbles — APPROVED** [review]

**Pull request merged**